### PR TITLE
Maven dep

### LINKED ISSUE #159: Support distribution of binaries
We have need for `mvn` or `mvnd` to be distributed on a build layer.  This makes a maven binary available for subsequent buildpacks to use.  In this scenario, the source project does not necessarily contain a `pom.xml`.

## Describe the Enhancement

The enhancement allows the maven buildpack to operate as a [distribution buildpack](https://github.com/paketo-buildpacks/community/blob/main/STYLEGUIDE.md#stack-choice).  Subsequent buildpacks can use the maven binary to perform maven operations.

## Possible Solution

The maven buildpack could succeed to detect if `BP_MAVEN_COMMAND` is set to either `mvn` or `mvnd`.  During the build phase, if `BP_MAVEN_COMMAND` is set to either `mvn` or `mvnd` then the corresponding binary is contributed as a build layer.

A buildpack that requires the maven buildpack might be expected to explicitly set a flag to disable running the maven command.  The default operation of the buildpack, in the absence of the flag, would continue to run the build command.

```go
Requires: []libcnb.BuildPlanRequire{
	{
		Name: maven.PlanEntryMaven,
		Metadata: map[string]interface{}{
			"RunBuild": false,
		},
	},
}
```

The current cache layer on which `mvn`/`mvd` are distributed would be made available as a build and cache layer.

## Motivation

Our motivation is to support injection of Jars into an application image in a non-JVM source project.  For example, python is a common language for data science projects and data science projects in python often use PySpark for distributed computation.  Such use-cases commonly need to install a single Jar file, or a small number of Jar files, and make them available in the CLASSPATH.  The set of Jars to install depend on the dependencies used in the Python project.   Application developers do not want to manage both a python build system and a maven/gradle/sbt build system.  Supporting a "distribution only" operation in the maven buildpack allows us to provide a subsequent buildpack to detect a PySpark project and inject appropriate Jar files.

Given this motivation, a data-science oriented build image might set `BP_MAVEN_COMMAND=mvn` by default.  The buildpack that contributes the Jars would depend on the maven buildpack and set `RunBuild: false`.  This moves the problem of managing dependencies in a `pom.xml` from the data-scientist to the platform operator.

### LINKED ISSUE #184: Document Include/Exclude files
With libbs 1.16.0, there is now support for including and excluding files. This needs to be documented in the Readme and also Paketo docs.

It needs to be documented for the other four libbs based Buildpacks as well.

### PR DESCRIPTION
**Pull request by dmikusa**
### Changes to Detection:

1. By default, we now pass
2. By default, we provide three possible plans all of which only provide (one provides just maven, one provides just a built app, one provides maven and a built app)
3. If a `pom.xml` file exists, then we add a requires (jdk, syft, maven).

### Changes to Build:

Based on what is in the buildplan, the buildpack can now optionally install and optionally build.

If `maven` is in the buildplan, then the buildpack will attempt to install or validate that Maven is installed and generate a command to run Maven.

This can be done in four ways:

1. Install Maven Daemon
2. Install Maven
3. Validate Maven Wrapper
4. Validate Maven on $PATH

If `jvm-application-package` is in the buildplan, then the buildpack will attempt to perform a build with Maven. This uses the command from the previously installed or validated Maven.

## Use Cases

This enables four scenarios:

1. The buildpack installs & runs Maven (or Maven Daemon)
2. The buildpack runs Maven w/Maven Wrapper
3. The buildpack just installs Maven (or Maven Daemon)
4. The buildpack doesn't install Maven or use Maven Wrapper, but just runs Maven (Maven is provided through other means)

For 1.) & 2.) are the classic cases we've been supporting.
For 3.) & 4.) are the new cases that have been requested.

Resolves #159 
Resolves #184

### CODE REVIEW

**Comment by dmikusa** [comment]
@AidanDelaney - FYI

**Comment by AidanDelaney** [comment]
This looks really good.  The `maven_manager` abstraction feels like the right separation of concerns.

File: maven/detect_test.go [hunk 1]
```diff
@@ -48,9 +48,32 @@ func testDetect(t *testing.T, context spec.G, it spec.S) {
 		Expect(os.RemoveAll(ctx.Application.Path)).To(Succeed())
 	})
 
-	it("fails without pom.xml", func() {
+	it("only provides with pom.xml", func() {
```
Comment (pivotal-david-osullivan) [inline maven/detect_test.go:51]:
```suggestion
	it("only provides without pom.xml", func() {
```
Should this read without?

**Comment by pivotal-david-osullivan** [comment]
> By default, we provide three possible plans all of which only provide

Don't we always still require a JDK?

> & 2. are the classic cases we've been supporting.   -- & 4. are the new cases that have been requested.

I think auto numbering messed these points up, not a big deal but it's a very useful summary so good to have it correct if so

**Comment by dmikusa** [comment]
> > By default, we provide three possible plans all of which only provide
> 
> Don't we always still require a JDK?

Right now, we contribute three plans always:

1. Provides `maven` + Requires `jdk`
2. Provides `jvm-application-package`
3. Provides `jvm-application-package` + Provides `maven`.

The thought is that for the first option, some other buildpack would need to require Maven. Thus, we require `jdk` so the other buildpack doesn't need to do that.

For two and three, the thought is that you need a `pom.xml` file & when a `pom.xml` is found, we automatically require `jdk`, `syft` and `maven` for those build plans. My understanding is that with build plans, the first one that matches will be picked so if a builder extension requires `jvm-application-package` and also meets the `jdk`, `syft` and `maven` requirements then it would get picked first. If not, then the default buildack option would be pick (i.e. option three, install & build).

Does that sound good?

> 
> > & 2. are the classic cases we've been supporting.   -- & 4. are the new cases that have been requested.
> 
> I think auto numbering messed these points up, not a big deal but it's a very useful summary so good to have it correct if so

Fixed.

File: maven/detect_test.go (hunk 1 above)
Comment (dmikusa) [reply maven/detect_test.go]: I pushed a change to address this. There were some other minor issues with the detect test. Good catch. Thanks

**Comment by pivotal-david-osullivan** [comment]
Yes it makes sense 👍  

My first point wasn't very clear, I just meant to clarify whether plan 1 makes this wording untrue: 'all of which only provide' since there is always a Requires for #1.

Looks good though and the extra tests make the scenarios easier to follow now 🎉

**Review by pivotal-david-osullivan — APPROVED** [review]

**Comment by pbusko** [comment]
@dmikusa this change broke the pipelines on our end due to the fact that the maven buildpack now detects always, without requiring JDK to be installed. We're building images from a pre-compiled jar file, and now we get the following error with the lates version of java buildpack:

```console
$ pack build --builder paketobuildpacks/builder:base --buildpack gcr.io/paketo-buildpacks/java:latest --path target/demo-1.0.0.jar test-image
Previous image with name "test-image" not found
===> DETECTING
7 of 24 buildpacks participating
paketo-buildpacks/ca-certificates   3.4.0
paketo-buildpacks/bellsoft-liberica 9.7.0
paketo-buildpacks/syft              1.20.0
paketo-buildpacks/maven             6.9.0
paketo-buildpacks/executable-jar    6.5.0
paketo-buildpacks/dist-zip          5.4.0
paketo-buildpacks/spring-boot       5.19.0
===> RESTORING
===> BUILDING

Paketo Buildpack for CA Certificates 3.4.0
  https://github.com/paketo-buildpacks/ca-certificates
  Launch Helper: Contributing to layer
    Creating /layers/paketo-buildpacks_ca-certificates/helper/exec.d/ca-certificates-helper

Paketo Buildpack for BellSoft Liberica 9.7.0
  https://github.com/paketo-buildpacks/bellsoft-liberica
  Build Configuration:
    $BP_JVM_JLINK_ARGS           --no-man-pages --no-header-files --strip-debug --compress=1  configure custom link arguments (--output must be omitted)
    $BP_JVM_JLINK_ENABLED        false                                                        enables running jlink tool to generate custom JRE
    $BP_JVM_TYPE                 JRE                                                          the JVM type - JDK or JRE
    $BP_JVM_VERSION              11                                                           the Java version
  Launch Configuration:
    $BPL_DEBUG_ENABLED           false                                                        enables Java remote debugging support
    $BPL_DEBUG_PORT              8000                                                         configure the remote debugging port
    $BPL_DEBUG_SUSPEND           false                                                        configure whether to suspend execution until a debugger has attached
    $BPL_HEAP_DUMP_PATH                                                                       write heap dumps on error to this path
    $BPL_JAVA_NMT_ENABLED        true                                                         enables Java Native Memory Tracking (NMT)
    $BPL_JAVA_NMT_LEVEL          summary                                                      configure level of NMT, summary or detail
    $BPL_JFR_ARGS                                                                             configure custom Java Flight Recording (JFR) arguments
    $BPL_JFR_ENABLED             false                                                        enables Java Flight Recording (JFR)
    $BPL_JMX_ENABLED             false                                                        enables Java Management Extensions (JMX)
    $BPL_JMX_PORT                5000                                                         configure the JMX port
    $BPL_JVM_HEAD_ROOM           0                                                            the headroom in memory calculation
    $BPL_JVM_LOADED_CLASS_COUNT  35% of classes                                               the number of loaded classes in memory calculation
    $BPL_JVM_THREAD_COUNT        250                                                          the number of threads in memory calculation
    $JAVA_TOOL_OPTIONS                                                                        the JVM launch flags
    Using Java version 11 extracted from MANIFEST.MF
  BellSoft Liberica JRE 11.0.16: Contributing to layer
    Downloading from https://github.com/bell-sw/Liberica/releases/download/11.0.16.1+1/bellsoft-jre11.0.16.1+1-linux-amd64.tar.gz
    Verifying checksum
    Expanding to /layers/paketo-buildpacks_bellsoft-liberica/jre
    Adding 127 container CA certificates to JVM truststore
    Writing env.launch/BPI_APPLICATION_PATH.default
    Writing env.launch/BPI_JVM_CACERTS.default
    Writing env.launch/BPI_JVM_CLASS_COUNT.default
    Writing env.launch/BPI_JVM_SECURITY_PROVIDERS.default
    Writing env.launch/JAVA_HOME.default
    Writing env.launch/JAVA_TOOL_OPTIONS.append
    Writing env.launch/JAVA_TOOL_OPTIONS.delim
    Writing env.launch/MALLOC_ARENA_MAX.default
  Launch Helper: Contributing to layer
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/active-processor-count
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/java-opts
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/jvm-heap
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/link-local-dns
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/memory-calculator
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/security-providers-configurer
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/jmx
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/jfr
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/security-providers-classpath-9
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/debug-9
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/nmt
    Creating /layers/paketo-buildpacks_bellsoft-liberica/helper/exec.d/openssl-certificate-loader
  Java Security Properties: Contributing to layer
    Writing env.launch/JAVA_SECURITY_PROPERTIES.default
    Writing env.launch/JAVA_TOOL_OPTIONS.append
    Writing env.launch/JAVA_TOOL_OPTIONS.delim

Paketo Buildpack for Syft 1.20.0
  https://github.com/paketo-buildpacks/syft
    Downloading from https://github.com/anchore/syft/releases/download/v0.57.0/syft_0.57.0_linux_amd64.tar.gz
    Verifying checksum
    Writing env.build/SYFT_CHECK_FOR_APP_UPDATE.default

Paketo Buildpack for Maven 6.9.0
  https://github.com/paketo-buildpacks/maven
  Build Configuration:
    $BP_MAVEN_BUILD_ARGUMENTS  -Dmaven.test.skip=true --no-transfer-progress package  the arguments to pass to Maven
    $BP_MAVEN_BUILT_ARTIFACT   target/*.[ejw]ar                                       the built application artifact explicitly.  Supersedes $BP_MAVEN_BUILT_MODULE
    $BP_MAVEN_BUILT_MODULE                                                            the module to find application artifact in
    $BP_MAVEN_DAEMON_ENABLED   false                                                  use maven daemon
    $BP_MAVEN_POM_FILE         pom.xml                                                the location of the main pom.xml file, relative to the application root
    $BP_MAVEN_SETTINGS_PATH                                                           the path to a Maven settings file

Paketo Buildpack for Maven 6.9.0
  unable to create application layer
  failed to generate expected metadata
  unable to determine java version
  error executing 'javac -version':
   Combined Output: :
  unable to start PTY
  exec: "javac": executable file not found in $PATH
ERROR: failed to build: exit status 1
ERROR: failed to build: executing lifecycle: failed with status code: 51
```

**Comment by dmikusa** [comment]
Thanks for the report. I thought we'd accounted for that case, but clearly not. We're going to roll this back temporarily. We need to add a bit more logic to the detect.

**Comment by dmikusa** [comment]
@pbusko Sorry for the issue, I have PR #189 which should correct this. 

If you have a minute to test & give us some feedback, I have a docker image up with this fix. `pack build <app-image> --buildpack dmikusa/paketo-java`. That'll pull from Docker Hub, it's a build of the standard Paketo Java buildpack but with an advanced build including PR #189.

**Comment by pbusko** [comment]
@dmikusa sorry for late reply, I can confirm that when using `dmikusa/paketo-java` buildpack, the issue no longer occurs. Maven buildpack doesn't detect anymore.

**Comment by dmikusa** [comment]
Excellent! thanks!

**Pull request merged**